### DISCUSSION TITLE
Using client extensions to return a translation?

### DISCUSSION THREAD

**Discussion by notflip** [prisma/prisma#20236] [Q&A]
I have a Category and CategoryTranslation table, I'm wondering if anyone knows how to use the client extensions to only return a CategoryTranslation for a given locale 'en'? Has anyone achieved something similar?

**Comment by ludralph** [comment]
Hi @notflip 👋 

Let's say we have these models:

```prisma
model Category {
  id            Int               @id @default(autoincrement())
  translations  CategoryTranslation[]
}

model CategoryTranslation {
  id            Int               @id @default(autoincrement())
  locale        String
  name          String
  category      Category          @relation(fields: [categoryId], references: [id])
  categoryId    Int
}
```

We can write a Prisma Client extension to return the translation for a given locale:

```js
const prisma = new PrismaClient().$extends({
  query: {
    category: {
      $allOperations({ args, query }) {
        // Refine the type - methods other than `create` accept a `where` clause
        const refinedArgs = args ;

        // Augment the `where` clause with `translations` filtered by the given locale
        const augmentedArgs = {
          ...refinedArgs,
          include: {
            translations: {
              where: { locale: "en" },
            },
          },
        };

        return query(augmentedArgs);
      },
    },
  },
});
```

We can then query like so:

```js
const categories = await prisma.category.findMany()
```
This will produce an output of the form:

```js
[
  {
    id: 1,
    translations: [ { id: 1, locale: 'en', name: 'Category Name', categoryId: 1 } ]
  },
  { id: 2, translations: [] },
  {
    id: 3,
    translations: [ { id: 3, locale: 'en', name: 'Category Name', categoryId: 3 } ]
  },
  {
    id: 4,
    translations: [ { id: 4, locale: 'en', name: 'Category Name', categoryId: 4 } ]
  }
]
```

Originally the categories array with the translations included looked like this:
```js
[
  {
    id: 1,
    translations: [ { id: 1, locale: 'en', name: 'Category Name', categoryId: 1 } ]
  },
  {
    id: 2,
    translations: [ { id: 2, locale: 'de', name: 'Category Name', categoryId: 2 } ]
  },
  {
    id: 3,
    translations: [ { id: 3, locale: 'en', name: 'Category Name', categoryId: 3 } ]
  },
  {
    id: 4,
    translations: [ { id: 4, locale: 'en', name: 'Category Name', categoryId: 4 } ]
  }
]
```

Notice that the `translations` with category id 2 returns an empty array `[ ]` after we applied the extension we created. 

Also returning only for the `de` locale gives the output
```js
[
  { id: 1, translations: [] },
  {
    id: 2,
    translations: [ { id: 2, locale: 'de', name: 'Category Name', categoryId: 2 } ]
  },
  { id: 3, translations: [] },
  { id: 4, translations: [] }
]
```

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by notflip** [reply]
Thanks for writing this out @ludralph The only issue I have now is that these client extensions are initiated on app boot I suppose? What if I want the language to depend on a session variable that can change. Is it possible to do that?

**Reply by ludralph** [reply]
Hi @notflip 

One solution I can think of is to have multiple Prisma Client Instances.
You could have extensions similar to this:
```js
const prismaEn = new PrismaClient().$extends({
  query: {
    category: {
      $allOperations({ args, query }) {
        // Refine the type - methods other than `create` accept a `where` clause
        const refinedArgs = args ;

        // Augment the `where` clause with `translations` filtered by the given locale
        const augmentedArgs = {
          ...refinedArgs,
          include: {
            translations: {
              where: { locale: "en" },
            },
          },
        };

        return query(augmentedArgs);
      },
    },
  },
});
```
The above one is for English, you could have another Prisma Client (with extension) for another language (like french) 
```js
const prismaFr = new PrismaClient().$extends({
  query: {
    category: {
      $allOperations({ args, query }) {
        // Refine the type - methods other than `create` accept a `where` clause
        const refinedArgs = args ;

        // Augment the `where` clause with `translations` filtered by the given locale
        const augmentedArgs = {
          ...refinedArgs,
          include: {
            translations: {
              where: { locale: "fr" },
            },
          },
        };

        return query(augmentedArgs);
      },
    },
  },
});
```

And when you want to switch a language dynamically, you could switch Prisma Client Instance as well.

```js
if(language == "en"){
 // use prismaEn
} else if(language == "fr"){
 // use prismaFr
}
```

You can also abstract the extension logic into a function and easily pass the locale as an argument. Using multiple Prisma clients would take up more database connections.

**Reply by ludralph** [reply]
A recommendation from the engineering team is that you use `$extends` per session rather than whole separate instances with your connection pool for each language.

**Reply by notflip** [reply]
@ludralph How do you mean per session? Do you have an example of this? Thanks for getting back to me.

**Reply by BjoernRave** [reply]
I think sth like this is what he was referring to: 
```javascript
//context.ts

/**
 * This is the actual context you'll use in your router
 * @link https://trpc.io/docs/context
 **/
export const createContext = async (opts: CreateNextContextOptions) => {
  const user = await getUser(opts.req)

  const localizedPrisma = prisma.$extends({
    query: {
      $allOperations: async ({ query, args }) => {
        const result = await query(args)

        if (Array.isArray(result)) {
          return result.map((r) => localizeModels({ r, lang: user.language }))
        }

        return localizeModels({ r: result, lang: user.language })
      },
    },
  })

  return {
    user,
    prisma: localizedPrisma,
  }
}

export type Context = inferAsyncReturnType<typeof createContext>
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
